miri: printing with full mir doesn't work because of pointer aligning inside memrchr

https://github.com/rust-lang/rust/blob/13e07a4e180e964717a0f71f0fc40260bab7d84a/src/libcore/slice/memchr.rs#L106

```
error: a raw memory access tried to access part of a pointer value as raw bytes

error[E0080]: constant evaluation error
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/slice/memchr.rs:106:21
    |
106 |     let end_align = (ptr as usize + len) & (usize_bytes - 1);
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ a raw memory access tried to access part of a pointer value as raw bytes
    |
note: inside call to `core::slice::memchr::memrchr`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libstd/sys/unix/memchr.rs:53:9
    |
53  |         ::core::slice::memchr::memrchr(needle, haystack)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `std::sys::unix::memchr::memrchr::memrchr_specific`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libstd/sys/unix/memchr.rs:56:5
    |
56  |     memrchr_specific(needle, haystack)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `std::sys::unix::memchr::memrchr`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libstd/memchr.rs:55:5
    |
55  |     ::sys::memchr::memrchr(needle, haystack)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `std::memchr::memrchr`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libstd/io/buffered.rs:905:23
    |
905 |         let i = match memchr::memrchr(b'/n', buf) {
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `<std::io::LineWriter<W> as std::io::Write><std::io::stdio::Maybe<std::io::stdio::StdoutRaw>>::write`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libstd/io/stdio.rs:465:9
    |
465 |         self.inner.borrow_mut().write(buf)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `<std::io::StdoutLock<'a> as std::io::Write>::write`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libstd/io/mod.rs:1111:19
    |
1111|             match self.write(buf) {
    |                   ^^^^^^^^^^^^^^^
[...]
```

--- a/src/libcore/slice/memchr.rs
+++ b/src/libcore/slice/memchr.rs
@@ -102,8 +102,24 @@ pub fn memrchr(x: u8, text: &[u8]) -> Option<usize> {
     let ptr = text.as_ptr();
     let usize_bytes = mem::size_of::<usize>();
 
+    // a version of align_offset that says how much must be *subtracted*
+    // from a pointer to be aligned.
+    #[inline(always)]
+    fn align_offset_down(ptr: *const u8, align: usize) -> usize {
+        let align_offset = ptr.align_offset(align);
+        if align_offset > align {
+            // Not possible to align
+            usize::max_value()
+        } else if align_offset == 0 {
+            0
+        } else {
+            // E.g. if align=8 and we have to add 1, then we can also subtract 7.
+            align - align_offset
+        }
+    }
+
     // search to an aligned boundary
-    let end_align = (ptr as usize + len) & (usize_bytes - 1);
+    let end_align = align_offset_down(unsafe { ptr.offset(len as isize) }, usize_bytes);
     let mut offset;
     if end_align > 0 {
         offset = if end_align >= len { 0 } else { len - end_align };

--- a/src/libcore/slice/memchr.rs
+++ b/src/libcore/slice/memchr.rs
@@ -102,32 +102,13 @@ pub fn memrchr(x: u8, text: &[u8]) -> Option<usize> {
     let ptr = text.as_ptr();
     let usize_bytes = mem::size_of::<usize>();
 
-    // a version of align_offset that says how much must be *subtracted*
-    // from a pointer to be aligned.
-    #[inline(always)]
-    fn align_offset_down(ptr: *const u8, align: usize) -> usize {
-        let align_offset = ptr.align_offset(align);
-        if align_offset > align {
-            // Not possible to align
-            usize::max_value()
-        } else if align_offset == 0 {
-            0
-        } else {
-            // E.g. if align=8 and we have to add 1, then we can also subtract 7.
-            align - align_offset
-        }
-    }
-
-    // search to an aligned boundary
-    let end_align = align_offset_down(unsafe { ptr.offset(len as isize) }, usize_bytes);
-    let mut offset;
-    if end_align > 0 {
-        offset = if end_align >= len { 0 } else { len - end_align };
-        if let Some(index) = text[offset..].iter().rposition(|elt| *elt == x) {
-            return Some(offset + index);
-        }
-    } else {
-        offset = len;
+    let mut offset = {
+        // We call this just to obtain the length of the suffix
+        let (_, _, suffix) = unsafe { text.align_to::<usize>() };
+        len - suffix.len()
+    };
+    if let Some(index) = text[offset..].iter().rposition(|elt| *elt == x) {
+        return Some(offset + index);
     }
 
     // search the body of the text

